### ISSUE TITLE
parallel::shared::Triangulation should exchange refinement flags

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#5359]
If I understand the implementation of `parallel::shared::Triangulation::execute_coarsening_and_refinement` correctly, i.e.,
```
    template <int dim, int spacedim>
    void
    Triangulation<dim,spacedim>::execute_coarsening_and_refinement ()
    {
      dealii::Triangulation<dim,spacedim>::execute_coarsening_and_refinement ();
      partition();
      this->update_number_cache ();
    }
```
then this function relies on the fact that the refinement flags are set in the same way on all participating processors. This invites bugs where someone sets these flags differently on one processor than another (for example, only setting flags on the locally owned cells), and the triangulations diverge.

As a first step, we should add an assert that after refinement all processors have the same number of cells. But the real solution is if every processor takes the flags from its own locally owned cells and merges them with those from the other processors. This takes away the possibility for things getting out of sync.

**Comment by davydden** [comment]
> this function relies on the fact that the refinement flags are set in the same way on all participating processors

correct. I think ideally there should be MPI exchange of refinement data among cells so that each process takes refinement data for not locally owned cells from other MPI cores.

**Comment by tjhei** [comment]
I was first thinking that a typical workflow would be like in the distributed case: estimate error for own cells, communicate flags to all other processors, then refine.

Of course this doesn't scale in parallel, so we might as well copy things around at different steps:
1) copy the solution vector to each processor, run the estimator everywhere.
2) copy the estimated errors around. This is what we do in step-18 for example.
3) copy the refinement flags.

I guess 3) would be the most robust solution? If yes, step-18 should be adapted.

I already have a few basic tests based on the mailing list discussion that I will create a PR for later this week.

**Comment by bangerth** [comment]
Thanks for dealing with this!

I agree that option 3 is the most robust. In fact, that's the only thing the triangulation can do -- it doesn't, for example, have access to the solution vector (or the vector of error indicators, for that matter), and users may have adjusted refinement flags after calling `GridRefinement` as well.

**Comment by davydden** [comment]
I also think (3) is the right way to do it here

**Comment by davydden** [comment]
For (1) you would communicate far more information with MPI. Communicating single bool flag per cell should be the least amount of data with p::s::Tria
